File: raspberryPi/touchIt_SET_ADD.py
#****************************************************************************************************
# touchIt_SET_ADD.py
#
# Get started with touchIt on Raspberry Pi! Go to www.probots.co.uk for more information.
#
# Configures the touchIt slave address. Once changed you'll have to edit TCHADD1!
#
# Demonstrates: "0x00 - Get Version", "0x02 - Set Address"
#
# Tested on Raspberry Pi Zero. Use "sudo apt-get install i2c-tools" (if necessary) and
#                                  "sudo i2cdetect -y 1" to check it is on the bus (device 0x70)
#
# If you get "no module named smbus", you'll need to "sudo apt-get install python-smbus" also ;)
#
# Written in Python 2.7.9
#
# P. Coates, ProBots, 17/09/16
# Free to use, modify, share!
#***************************************************************************************************

# Needed for I2C + GPIO
import time
import smbus
import struct
import RPi.GPIO as GPIO

DEVICE_BUS = 1                      # using pins 3 and 5 on Raspberry PI Zero
intPin = 17                         # using BCM 17 for touchIt interrupt output (pin 11)
bus = smbus.SMBus(DEVICE_BUS)

# touchIt commands
GET_VER = 0
SET_CFG = 1
SET_ADD = 2
GET_POS = 3
GET_TCH = 4

# touchIt config
XMOVE = 0x01
YMOVE = 0x02
TMOVE = 0x04
TAP   = 0x08

# touchIt address
TCHADD1 = 0x70

# touchIt new address (1-127)
TCHADD2 = 0x72

crc = 0
c = 0

#**************************************************
# Get the touchIt Firmware Version
#**************************************************
def getVersion():

    # Generate crc
    data = []
    data.append('#')                # Start
    data.append(chr(GET_VER))       # Command
    data.append(chr(0))             # Data Length

    crc = 0
    for x in range(len(data)):
        crc ^= ord(data[x])         # Calculated crc

    # Send command
    # note: no target register so we're sticking our first byte '#' in its place
    bus.write_i2c_block_data(TCHADD1, ord('#'), [GET_VER, 0, crc])
    time.sleep(0.6)                 # prevents smbus missing return data

    # Receive confirmation
    data = bus.read_i2c_block_data(TCHADD1, 0, 6)
    
    # Validate returned crc
    if data[0] == ord('$'):
        crc = 0

        for c in range(5):          # Calculate crc from incoming data (this command should return 5 bytes)
            crc ^= data[c]

        if crc == data[5]:          # check against received crc
            print "touchIt FW Ver: " + str(data[3]) + '.' + str(data[4])

#**************************************************
# Set a new touchIt address
#**************************************************
def setAddress(newAdd):

    # Generate crc
    data = []
    data.append('#')                # Start
    data.append(chr(SET_ADD))       # Command
    data.append(chr(1))             # Data Length
    data.append(chr(newAdd))        # Data

    crc = 0
    for x in range(len(data)):
        crc ^= ord(data[x])         # Calculated crc

    # Send command (using current address)
    bus.write_i2c_block_data(TCHADD1, ord('#'), [SET_ADD, 1, newAdd, crc])
    time.sleep(0.6)
    
    # No confirmation, address has changed

#**************************************************
# Configure the pi and touchIt
#**************************************************

# Configure input for reading the touchIt interrupt state
GPIO.setmode(GPIO.BCM)              # Use Broadcom pin-numbering scheme
GPIO.setup(intPin, GPIO.IN)

# Get touchIt FW version
getVersion()

# Interrupt on X/Y movement
setAddress(TCHADD2)

print "touchIt I2C address changed from %s to %s" % (hex(TCHADD1), hex(TCHADD2))

File: raspberryPi/touchIt_SET_ADD.c
/****************************************************************************************************
* touchIt_SET_ADD.c
*
* Get started with touchIt on Raspberry Pi! Go to www.probots.co.uk for more information.
*
* Configures the touchIt slave address. Once changed you'll have to edit TCHADD1!
*
* Demonstrates: "0x00 - Get Version", "0x02 - Set Address"
*
* Tested on Raspberry Pi Zero. Use "sudo apt-get install i2c-tools" (if necessary) and
*                                  "sudo i2cdetect -y 1" to check it is on the bus (device 0x70)
*
* If you get "no module named smbus", you'll need to "sudo apt-get install python-smbus" also ;)
*
* Written in nano 2.2.6
* Compiled using "gcc -o touchIt_SET_ADD touchIt_SET_ADD.c"
*
* P. Coates, ProBots, 17/09/16
* Free to use, modify, share!
***************************************************************************************************/

#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/i2c-dev.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>

// touchIt commands
#define GET_VER	0x00
#define SET_CFG	0x01
#define SET_ADD	0x02
#define GET_POS	0x03
#define GET_TCH	0x04

// touchIt config
#define XMOVE	0x01
#define YMOVE 	0x02
#define TMOVE	0x04
#define TAP	0x08

// touchIt address
#define TCHADD1 0x70

// touchIt new address (1-127)
#define TCHADD2 0x72

// GPIO params
#define BUFMAX 	3
#define DIRMAX	35
#define IN	0
#define OUT	1
#define LOW	0
#define HIGH 	1

int file_i2c;
int file_gpio;
int length;
unsigned char buffer[60] = {0};

unsigned char data[10];
unsigned char crc = 0, c = 0;
unsigned char intPin = 17;

/**************************************************
* Get the touchIt Firmware Version
**************************************************/
void getVersion(void)
{
	unsigned char len = 4;

	// Generate crc
	crc  = data[0] = '#';			// Start
	crc ^= data[1] = GET_VER;		// Command
	crc ^= data[2] = 0x00;			// Data Length
	data[3] = crc;

	// Send command
	if(write(file_i2c, data, len) != len)
	{
		printf("getVersion(): Failed to write to the i2c bus.\n");
		return;
	}

	// Receive confirmation
	len = 6;
	if(read(file_i2c, data, len) != len)
	{
		printf("getVersion(): Failed to read from the i2c bus.\n");
	}
	else
	{
		if(data[0] == '$')
		{
			crc = 0;
			for(c=0;c<5;c++)
			{
				crc ^= data[c];
			}

			if(crc == data[5])
			{
				printf("touchIt FW Ver: %i.%i\n", data[3], data[4]);
			}
		}
	}
}

/**************************************************
* Set a new touchIt address
**************************************************/
void setAddress(unsigned char newAdd)
{
	unsigned char len = 5;

	// Generate crc
	crc  = data[0] = '#';			// Start
	crc ^= data[1] = SET_ADD;		// Command
	crc ^= data[2] = 0x01;			// Data length
	crc ^= data[3] = newAdd;		// Data
	data[4] = crc;

	// Send command
	if(write(file_i2c, data, len) != len)
	{
		printf("setAddress(): Failed to write to the i2c bus\n");
		return;
	}

	// No confirmation, address has changed
}

/**************************************************
* GPIO read function for interrupt polling
**************************************************/
unsigned char gpioRead(void)
{
	unsigned char val[3];

	read(file_gpio, val, 3);

	lseek(file_gpio, 0, SEEK_SET);	// restore file pointer for next read

	return(atoi(val));
}

/**************************************************
* Exit method - clean up file handles etc
**************************************************/
void exitHandler(int s)
{
	if(s == SIGINT)
	{
		close(file_gpio);
		close(file_i2c);

		exit(1);
	}
}

/**************************************************
* Configure the Pi / touchIt
**************************************************/
void setup()
{
	unsigned char len = 0;
	unsigned char dir = IN;
	static const char dir_str[] = "in\0out";

	signal (SIGINT, exitHandler);

	// Open the I2C bus (enable file access)
	char *filename = (char *)"/dev/i2c-1";
	if((file_i2c = open(filename, O_RDWR)) < 0)
	{
		printf("setup(): Failed to open the i2c bus\n");
		return;
	}

	if(ioctl(file_i2c, I2C_SLAVE, TCHADD1) < 0)
	{
		printf("Failed to acquire bus access and/or talk to slave\n");
		return;
	}

	// Enable GPIO to read touchIt 'int' pin
	if((file_gpio = open("/sys/class/gpio/export", O_WRONLY)) < 0)
	{
		printf("setup(): Failed to open export for writing\n");
		return;
	}
	else
	{
		len = snprintf(buffer, BUFMAX, "%d", intPin);
		write(file_gpio, buffer, len);
		close(file_gpio);
	}

	// Configure GPIO Direction as input
	snprintf(buffer, DIRMAX, "/sys/class/gpio/gpio%d/direction", intPin);
	if((file_gpio = open(buffer, O_WRONLY)) < 0)
	{
		printf("setup(): Failed to open gpio direction for writing\n");
		return;
	}
	else
	{
		write(file_gpio, &dir_str[IN == dir ? 0:3], IN  == dir ? 2:3);
		close(file_gpio);
	}

	// Open gpio file for reading
	snprintf(buffer, DIRMAX, "/sys/class/gpio/gpio%d/value", intPin);
	if((file_gpio = open(buffer, O_RDONLY)) < 0)
	{
		printf("gpioRead(): failed to open gpio value for reading\n");
		return;
	}

	getVersion();

	setAddress(TCHADD2);

	printf("touchIt I2C address changed from %X to %X\n", TCHADD1, TCHADD2);
}

/**************************************************
* Do all the cool stuff
**************************************************/
int main()
{
	setup();

	while(1)
	{
	}
}

